File: auto-setup.js
#!/usr/bin/env node

/**
 * 🤖 AUTOMATED AZURE SETUP SCRIPT
 * This script will automatically:
 * 1. Retrieve Cosmos DB credentials
 * 2. Retrieve Storage Account credentials  
 * 3. Set environment variables in App Service
 * 4. Restart the app
 * 5. Verify everything works
 */

const { execSync } = require('child_process');
const fs = require('fs');

const colors = {
  reset: '\x1b[0m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  cyan: '\x1b[36m',
  blue: '\x1b[34m'
};

function log(color, message) {
  console.log(`${color}${message}${colors.reset}`);
}

function runCommand(command, silent = false) {
  try {
    const result = execSync(command, {
      encoding: 'utf-8',
      stdio: silent ? 'pipe' : 'inherit'
    }).trim();
    return result;
  } catch (error) {
    if (!silent) {
      log(colors.red, `❌ Command failed: ${command}`);
      log(colors.red, error.message);
    }
    return null;
  }
}

async function main() {
  log(colors.cyan, '\n═══════════════════════════════════════════════════════════');
  log(colors.cyan, '  🤖 AUTOMATED AZURE FILE MANAGER SETUP');
  log(colors.cyan, '═══════════════════════════════════════════════════════════\n');

  // Constants
  const SUBSCRIPTION_ID = 'ce176ab4-0474-47f5-bfe2-72e93937970f';
  const COSMOS_ACCOUNT = 'filemanagercosmos1234';
  const APP_SERVICE_NAME = 'file-manager-backend-app';
  const STORAGE_ACCOUNT_NAME = ''; // Will be found
  let RESOURCE_GROUP = 'file-manager-rg';

  try {
    // Step 1: Get Cosmos DB keys
    log(colors.yellow, '📍 Step 1: Retrieving Cosmos DB credentials...');
    
    let cosmosData = null;
    try {
      const cosmosJson = runCommand(
        `az cosmosdb list --subscription "${SUBSCRIPTION_ID}" --query "[?name=='${COSMOS_ACCOUNT}']" -o json`,
        true
      );
      if (cosmosJson) {
        cosmosData = JSON.parse(cosmosJson)[0];
        RESOURCE_GROUP = cosmosData.resourceGroup;
        log(colors.green, `✓ Cosmos DB found in resource group: ${RESOURCE_GROUP}`);
      }
    } catch (e) {
      log(colors.red, `❌ Could not parse Cosmos DB data`);
    }

    // Step 2: Get Cosmos DB keys
    log(colors.yellow, '📍 Step 2: Getting Cosmos DB keys...');
    
    let COSMOS_ENDPOINT = '';
    let COSMOS_KEY = '';
    
    try {
      const keysJson = runCommand(
        `az cosmosdb keys list --name "${COSMOS_ACCOUNT}" --resource-group "${RESOURCE_GROUP}" --type keys -o json`,
        true
      );
      
      if (keysJson) {
        const keys = JSON.parse(keysJson);
        COSMOS_KEY = keys.primaryMasterKey || keys.primaryKey;
        
        // Get endpoint
        const endpointJson = runCommand(
          `az cosmosdb show --name "${COSMOS_ACCOUNT}" --resource-group "${RESOURCE_GROUP}" -o json`,
          true
        );
        
        if (endpointJson) {
          const cosmosInfo = JSON.parse(endpointJson);
          COSMOS_ENDPOINT = cosmosInfo.documentEndpoint;
        }
        
        log(colors.green, '✓ Cosmos DB credentials retrieved');
        log(colors.green, `  - Endpoint: ${COSMOS_ENDPOINT.substring(0, 50)}...`);
        log(colors.green, `  - Key: ${COSMOS_KEY.substring(0, 20)}...`);
      }
    } catch (e) {
      log(colors.red, `❌ Failed to get Cosmos DB keys: ${e.message}`);
    }

    // Step 3: Get Storage Account connection string
    log(colors.yellow, '\n📍 Step 3: Retrieving Storage Account credentials...');
    
    let STORAGE_CONNECTION_STRING = '';
    
    try {
      // Find storage accounts
      const storageJson = runCommand(
        `az storage account list --resource-group "${RESOURCE_GROUP}" -o json`,
        true
      );
      
      if (storageJson) {
        const storageAccounts = JSON.parse(storageJson);
        if (storageAccounts.length > 0) {
          const storageAccount = storageAccounts[0];
          const storageName = storageAccount.name;
          
          // Get connection string
          const connStrJson = runCommand(
            `az storage account show-connection-string --name "${storageName}" --resource-group "${RESOURCE_GROUP}" -o json`,
            true
          );
          
          if (connStrJson) {
            const connInfo = JSON.parse(connStrJson);
            STORAGE_CONNECTION_STRING = connInfo.connectionString;
            log(colors.green, '✓ Storage Account credentials retrieved');
            log(colors.green, `  - Storage: ${storageName}`);
            log(colors.green, `  - Connection String: ${STORAGE_CONNECTION_STRING.substring(0, 50)}...`);
          }
        }
      }
    } catch (e) {
      log(colors.red, `❌ Failed to get Storage Account: ${e.message}`);
    }

    // Step 4: Set environment variables in App Service
    if (COSMOS_ENDPOINT && COSMOS_KEY && STORAGE_CONNECTION_STRING) {
      log(colors.yellow, '\n📍 Step 4: Setting environment variables in App Service...');
      
      const settings = [
        `COSMOS_ENDPOINT="${COSMOS_ENDPOINT}"`,
        `COSMOS_KEY="${COSMOS_KEY}"`,
        `AZURE_STORAGE_CONNECTION_STRING="${STORAGE_CONNECTION_STRING}"`,
        `CONTAINER_NAME="files"`,
        `COSMOS_DB_NAME="file-notes-db"`,
        `COSMOS_CONTAINER_NAME="files"`
      ].join(' ');
      
      try {
        runCommand(
          `az webapp config appsettings set --resource-group "${RESOURCE_GROUP}" --name "${APP_SERVICE_NAME}" --settings ${settings}`,
          false
        );
        log(colors.green, '✓ Environment variables set');
      } catch (e) {
        log(colors.red, `❌ Failed to set environment variables: ${e.message}`);
      }

      // Step 5: Restart App Service
      log(colors.yellow, '\n📍 Step 5: Restarting App Service...');
      
      try {
        runCommand(
          `az webapp restart --resource-group "${RESOURCE_GROUP}" --name "${APP_SERVICE_NAME}"`,
          false
        );
        log(colors.green, '✓ App Service restarted');
      } catch (e) {
        log(colors.red, `❌ Failed to restart App Service: ${e.message}`);
      }

      // Step 6: Wait for restart
      log(colors.yellow, '\n📍 Step 6: Waiting for app to fully restart (60 seconds)...');
      
      let progress = 0;
      const interval = setInterval(() => {
        progress += 5;
        process.stdout.write(`\r⏳ ${progress}s elapsed...`);
        if (progress >= 60) {
          clearInterval(interval);
          console.log('\n');
        }
      }, 5000);
      
      await new Promise(resolve => setTimeout(resolve, 60000));

      // Step 7: Verify setup
      log(colors.yellow, '📍 Step 7: Verifying setup...');
      
      try {
        const response = await fetch(
          'https://file-manager-backend-app.azurewebsites.net/api/files/diagnostics'
        );
        const data = await response.json();
        
        if (data.environment) {
          const allSet = Object.values(data.environment).every(v => v === '✓');
          if (allSet) {
            log(colors.green, '✅ ALL ENVIRONMENT VARIABLES ARE SET!');
            log(colors.green, '\nVerification Details:');
            Object.entries(data.environment).forEach(([key, val]) => {
              log(colors.green, `  ✓ ${key}: ${val}`);
            });
            log(colors.green, '\n✅ SETUP COMPLETE! You can now upload files! 🚀');
          } else {
            log(colors.red, '❌ Some environment variables are not set');
            log(colors.red, JSON.stringify(data.environment, null, 2));
          }
        }
      } catch (e) {
        log(colors.red, `⚠️  Could not verify setup: ${e.message}`);
        log(colors.yellow, 'But settings have been applied. Wait 1-2 more minutes and check:');
        log(colors.cyan, 'https://file-manager-backend-app.azurewebsites.net/api/files/diagnostics');
      }
    } else {
      log(colors.red, '\n❌ Could not retrieve all credentials. Aborting setup.');
      log(colors.yellow, 'Please check:');
      log(colors.yellow, `  - Cosmos DB Account: ${COSMOS_ACCOUNT}`);
      log(colors.yellow, `  - Resource Group: ${RESOURCE_GROUP}`);
      log(colors.yellow, `  - Storage Account exists in resource group`);
    }

  } catch (error) {
    log(colors.red, `\n❌ Setup failed: ${error.message}`);
    log(colors.red, error.stack);
  }

  log(colors.cyan, '\n═══════════════════════════════════════════════════════════\n');
}

main().catch(err => {
  log(colors.red, `Fatal error: ${err.message}`);
  process.exit(1);
});

File: auto_setup.py
#!/usr/bin/env python3
"""
🤖 Fully Automated Azure Setup Script
Automatically configures all environment variables and restarts the app
"""

import subprocess
import json
import time
import sys
from urllib.request import urlopen
from urllib.error import URLError

class Colors:
    CYAN = '\033[96m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    RED = '\033[91m'
    BLUE = '\033[94m'
    RESET = '\033[0m'

def print_header(msg):
    print(f"{Colors.CYAN}{'='*60}{Colors.RESET}")
    print(f"{Colors.CYAN}  {msg}{Colors.RESET}")
    print(f"{Colors.CYAN}{'='*60}{Colors.RESET}\n")

def print_step(step, msg):
    print(f"{Colors.YELLOW}📍 Step {step}: {msg}{Colors.RESET}")

def print_success(msg):
    print(f"{Colors.GREEN}✓ {msg}{Colors.RESET}")

def print_error(msg):
    print(f"{Colors.RED}❌ {msg}{Colors.RESET}")

def run_command(cmd, silent=False):
    """Run Azure CLI command and return output"""
    try:
        result = subprocess.run(
            cmd,
            shell=True,
            capture_output=True,
            text=True,
            timeout=30
        )
        if result.returncode != 0:
            if not silent:
                print_error(f"Command failed: {cmd}")
                print_error(result.stderr)
            return None
        return result.stdout.strip()
    except Exception as e:
        if not silent:
            print_error(f"Error running command: {e}")
        return None

def main():
    print_header("🤖 AUTOMATED AZURE FILE MANAGER SETUP")
    
    # Configuration
    SUBSCRIPTION_ID = "ce176ab4-0474-47f5-bfe2-72e93937970f"
    COSMOS_ACCOUNT = "filemanagercosmos1234"
    APP_SERVICE_NAME = "file-manager-backend-app"
    RESOURCE_GROUP = "file-manager-rg"
    
    print("Starting automated setup...\n")
    
    # ========================================
    # STEP 1: Get Cosmos DB Info
    # ========================================
    print_step(1, "Retrieving Cosmos DB credentials")
    
    cosmos_info_json = run_command(
        f'az cosmosdb show --name {COSMOS_ACCOUNT} --resource-group {RESOURCE_GROUP} -o json'
    )
    
    if not cosmos_info_json:
        print_error("Could not get Cosmos DB info")
        return False
    
    try:
        cosmos_info = json.loads(cosmos_info_json)
        COSMOS_ENDPOINT = cosmos_info['documentEndpoint']
        print_success(f"Cosmos DB Endpoint: {COSMOS_ENDPOINT[:50]}...")
    except Exception as e:
        print_error(f"Could not parse Cosmos DB info: {e}")
        return False
    
    # Get Cosmos DB keys
    keys_json = run_command(
        f'az cosmosdb keys list --name {COSMOS_ACCOUNT} --resource-group {RESOURCE_GROUP} --type keys -o json'
    )
    
    if not keys_json:
        print_error("Could not get Cosmos DB keys")
        return False
    
    try:
        keys_info = json.loads(keys_json)
        COSMOS_KEY = keys_info.get('primaryMasterKey') or keys_info.get('primaryKey')
        print_success(f"Cosmos DB Key: {COSMOS_KEY[:20]}...")
    except Exception as e:
        print_error(f"Could not parse Cosmos DB keys: {e}")
        return False
    
    # ========================================
    # STEP 2: Get Storage Account Info
    # ========================================
    print_step(2, "Retrieving Storage Account credentials")
    
    storage_list_json = run_command(
        f'az storage account list --resource-group {RESOURCE_GROUP} -o json'
    )
    
    if not storage_list_json:
        print_error("Could not list Storage Accounts")
        return False
    
    try:
        storage_accounts = json.loads(storage_list_json)
        if not storage_accounts:
            print_error("No Storage Accounts found in resource group")
            return False
        
        storage_account = storage_accounts[0]
        STORAGE_NAME = storage_account['name']
        print_success(f"Storage Account: {STORAGE_NAME}")
    except Exception as e:
        print_error(f"Could not parse Storage Accounts: {e}")
        return False
    
    # Get Storage Connection String
    storage_conn_json = run_command(
        f'az storage account show-connection-string --name {STORAGE_NAME} --resource-group {RESOURCE_GROUP} -o json'
    )
    
    if not storage_conn_json:
        print_error("Could not get Storage Connection String")
        return False
    
    try:
        storage_conn_info = json.loads(storage_conn_json)
        STORAGE_CONNECTION_STRING = storage_conn_info['connectionString']
        print_success(f"Storage Connection String: {STORAGE_CONNECTION_STRING[:50]}...")
    except Exception as e:
        print_error(f"Could not parse Storage Connection String: {e}")
        return False
    
    # ========================================
    # STEP 3: Set Environment Variables
    # ========================================
    print_step(3, "Setting environment variables in App Service")
    
    # Build the settings command
    settings_cmd = (
        f'az webapp config appsettings set '
        f'--resource-group {RESOURCE_GROUP} '
        f'--name {APP_SERVICE_NAME} '
        f'--settings '
        f'COSMOS_ENDPOINT="{COSMOS_ENDPOINT}" '
        f'COSMOS_KEY="{COSMOS_KEY}" '
        f'AZURE_STORAGE_CONNECTION_STRING="{STORAGE_CONNECTION_STRING}" '
        f'CONTAINER_NAME="files" '
        f'COSMOS_DB_NAME="file-notes-db" '
        f'COSMOS_CONTAINER_NAME="files"'
    )
    
    result = run_command(settings_cmd)
    if not result:
        print_error("Could not set environment variables")
        return False
    
    print_success("Environment variables set successfully")
    
    # ========================================
    # STEP 4: Restart App Service
    # ========================================
    print_step(4, "Restarting App Service")
    
    result = run_command(
        f'az webapp restart --resource-group {RESOURCE_GROUP} --name {APP_SERVICE_NAME}'
    )
    
    if not result and result is not None:
        print_error("Could not restart App Service")
        return False
    
    print_success("App Service restarting...")
    
    # ========================================
    # STEP 5: Wait and Verify
    # ========================================
    print_step(5, "Waiting for app to restart and verifying setup (60 seconds)")
    
    for i in range(0, 60, 5):
        sys.stdout.write(f"\r⏳ {i} seconds elapsed...")
        sys.stdout.flush()
        time.sleep(5)
    
    print("\r✓ Waiting complete                       ")
    print()
    
    print(f"{Colors.YELLOW}🔍 Verifying setup...{Colors.RESET}")
    
    max_attempts = 5
    for attempt in range(max_attempts):
        try:
            response = urlopen('https://file-manager-backend-app.azurewebsites.net/api/files/diagnostics', timeout=5)
            data_json = response.read().decode('utf-8')
            data = json.loads(data_json)
            
            if 'environment' in data:
                env = data['environment']
                all_set = all(v == '✓' for v in env.values())
                
                if all_set:
                    print(f"\n{Colors.GREEN}{'='*60}{Colors.RESET}")
                    print(f"{Colors.GREEN}✅  SETUP COMPLETE! SUCCESS!{Colors.RESET}")
                    print(f"{Colors.GREEN}{'='*60}{Colors.RESET}\n")
                    
                    print(f"{Colors.GREEN}📊 Verification Results:{Colors.RESET}")
                    for key, val in env.items():
                        print(f"  ✓ {key}: {val}")
                    
                    print(f"\n{Colors.GREEN}🎉 Your backend is now fully configured!{Colors.RESET}")
                    print(f"{Colors.GREEN}🚀 You can now upload files without errors!{Colors.RESET}")
                    print(f"\n{Colors.CYAN}🧪 Test it:{Colors.RESET}")
                    print(f"   1. Go to: https://file-manager-frontend-app.azurewebsites.net")
                    print(f"   2. Select a file and upload")
                    print(f"   3. File should upload successfully! ✅\n")
                    
                    print(f"{Colors.CYAN}{'='*60}{Colors.RESET}\n")
                    return True
        except Exception as e:
            if attempt < max_attempts - 1:
                print(f"⏳ Checking again... (Attempt {attempt+1}/{max_attempts})")
                time.sleep(10)
    
    print(f"\n{Colors.YELLOW}⚠️  Could not fully verify setup, but settings have been applied!{Colors.RESET}")
    print(f"{Colors.YELLOW}⏳ Wait 2-3 more minutes and check manually:{Colors.RESET}")
    print(f"{Colors.CYAN}   https://file-manager-backend-app.azurewebsites.net/api/files/diagnostics{Colors.RESET}\n")
    
    return True

if __name__ == '__main__':
    try:
        success = main()
        sys.exit(0 if success else 1)
    except Exception as e:
        print_error(f"Fatal error: {e}")
        sys.exit(1)
